src/seq/abif/reader.jl

```diff
@@ -0,0 +1,254 @@
+    return getindex(p, p.dirs[k]), k + 1
+end
+
+function Base.length(a::AbifReader)
+    return length(a.dirs)
+end
+
+function Base.checkbounds(p::AbifReader, k::Integer)
+    if 1 ≤ k ≤ length(p)
+        return true
+    end
+    throw(BoundsError(p, k))
+end
+
+function Base.getindex(a::AbifReader, k::Integer)
+    checkbounds(a, k)
+    return Dict([parse_data_tag(a, a.dirs[k])])
+end
+
+function Base.getindex(a::AbifReader, t::AbstractString)
+    tag = filter((x) -> x.name == t, a.dirs)
+
+    if isempty(tag)
+        throw(KeyError(t))
+    end
+
+    if length(tag) > 1
+        return Dict([parse_data_tag(a, b) for b in tag])
+     end
+
+    return Dict([parse_data_tag(a, first(tag))])
+end
+
+function Base.getindex(a::AbifReader, t::AbifDirEntry)
+    return Dict([parse_data_tag(a, t)])
+end
+
+function Base.getindex(a::AbifReader, t::Array{AbifDirEntry})
+    return Dict([parse_data_tag(a, k) for k in t])
+end
+
+# Returns all existing tags
+function tags(a::AbifReader)
+    return [tag for tag in a.dirs]
+end
+
+# Returns all existing tags by name
+function tags(a::AbifReader, t::AbstractString)
+    return [tag for tag in a.dirs if isequal(tag.name, t)]
+end
+
+# Returns the number of Tags by name
+function elements(a::AbifReader, t::AbstractString)
+    return length(tags(a, t))
+end
+
+# extract the header
+function read_abif_header(input::IO)
+    seekstart(input)
+    signature = read(input, 4)
+
+    if issupported(signature)
+        version = ntoh(first(read(input, Int16, 1)))
+        return parse_directory(input, position(input))
+    end
+end
+
+# extract all tags in file
+function read_abif_tags(input::IO, header::AbifDirEntry)
+    tags = AbifDirEntry[]
+    for index in collect(1:header.num_elements)
+        start = header.data_offset + index * header.element_size
+        tag = parse_directory(input, start)
+        push!(tags, tag)
+    end
+    return tags
+end
+
+# extract DirEtry
+function parse_directory(input::IO, pos::Int64)
+    seek(input, pos)
+
+    name         = String(read(input, 4))
+    number       = Int(ntoh(first(read(input, Int32, 1))))
+    element_type = Int(ntoh(first(read(input, UInt16, 1))))
+    element_size = Int(ntoh(first(read(input, Int16, 1))))
+    num_elements = Int(ntoh(first(read(input, Int32, 1))))
+    data_size    = Int(ntoh(first(read(input, Int32, 1))))
+
+    if data_size <= 4
+        data_offset = position(input)
+    else
+        data_offset = ntoh(first(read(input, Int32, 1)))
+    end
+
+    AbifDirEntry(name, number, element_type, element_size, num_elements, data_size, data_offset)
+end
+
+# read bytes according to the element type, other values are unsupported or legacy.
+function parse_data_tag(a::AbifReader, tag::AbifDirEntry)
+    if tag.data_offset > 0
+        seek(a.input, tag.data_offset)
+
+        if tag.element_type == 2
+            data = String(read(a.input, tag.data_size))
+
+            return format_data!(data, tag, elements(a, tag.name))
```

Comment:
If every branch finishes with `return format_data!(data, tag, elements(a, tag.name))`, it would be concise to place the return statement at the end of the function.